types/math/math.go

```diff
@@ -110,3 +110,37 @@ func SafeSub(res, x, y *apd.Decimal) error {
 
 	return nil
 }
+
+// SubBalance subtracts the value of y from x and returns the result with arbitrary precision.
+// Returns with ErrInsufficientFunds error if the result is negative.
+func SubBalance(x Dec, y Dec) (Dec, error) {
+	var z Dec
+	_, err := exactContext.Sub(&z.dec, &x.dec, &y.dec)
+	if err != nil {
+		return z, errors.Wrap(err, "decimal subtraction error")
+	}
+
+	if z.IsNegative() {
+		return z, errors.ErrInsufficientFunds
+	}
+
+	return z, nil
+}
+
+// AddBalance adds the value of x+y and returns the result with arbitrary precision.
+// Returns with ErrInvalidRequest error if either x or y is negative.
+func AddBalance(x Dec, y Dec) (Dec, error) {
+	var z Dec
+	_, err := exactContext.Add(&z.dec, &x.dec, &y.dec)
+	if err != nil {
+		return z, errors.Wrap(err, "decimal subtraction error")
+	}
+
+	if x.IsNegative() || y.IsNegative() {
```

Comment:
Let's do this before doing the actual addition to avoid an unnecessary operation!